doc: Add interop deployment documentation (#649)
TheThingsIndustries/lorawan-stack-docs: doc/content/getting-started/aws/ecs/interop/_index.md

---
title: "Interoperability"
description: "LoRaWAN Backend Interfaces"
weight: 3
---

{{% tts %}} exposes the Join Server API defined by LoRaWAN Backend Interfaces 1.0 and 1.1. This is used for interoperability between LoRaWAN networks. The functionality can be exposed either by the Join Server or by the Identity Server, but not both at once.

Join Server is capable of handling the join flow, that is `HomeNSReq`, `JoinReq` and `AppSKeyReq` requests. Identity Server can only answer to `HomeNSReq`. Typically, you should enable interoperability if you want to use device activations through Packet Broker. You can use Join Server interoperability if you want to expose the entire join flow to other LoRaWAN network servers.

# Join Server interoperability

## Deployment

### Step 1: deploy {{% tts %}}

> Deploy {{% tts %}} normally. For the `PeerRequesterAccountId` parameter in `1-1-vpc`, use the ID of the account where you want to deploy Crypto Server. Do not enable interop and assume that {{% tts %}} doesn't connect to the Crypto Server.

Reviewer: There is no need to document anything about Crypto Server at this point. It is part of TTSE but it is unusable by customers in its current shape and form.

Author: I think we should document this. This documentation isn't only for the customers, but also for ourselves. I mean, you spent hours looking up how Global Join Server actually works, so when we're doing a new setup, it's worth it to document it, so we don't run into the same problem again in the future.

Also, without the Crypto Server part, the documentation boils down to "TTS supports interop", so I can simply close this PR without merging.

Reviewer: OK

### Step 2: add certificates

For the server certificate, use the certbot task. It will request a certificate and put it into AWS Secrets Manager. If you do not want to use the certbot task, but prefer to provide your own certificate, look up the relevant secret created in template `4-1-secrets` and substitute the dummy value. Certificates are expected to be in PEM format.

> Client certificates are stored in the interop config bucket, created in template `2-4b-routing-s3`. In this bucket you need to place a file `config.yml`, which contains a dictionary `NetID -> filename_with_certificate_in_pem_format`. Example configuration:

Reviewer: ```suggestion
Client certificates are stored in the interop config bucket, created in template `2-4b-routing-s3`. In this bucket you need to place a file `config.yml`, which contains a map `<net-id>: <ca.pem>`. Example configuration:
```

> ```

Reviewer: ```suggestion
```yaml
```

### Step 6: encrypt key values for secure element parts

Go to the Crypto Server stack and encrypt secure element parts keys using the AWS KMS key from stack output `MicrochipPartsTableKey` and put them into the DynamoDB table from output `MicrochipPartsTable`. You can do it in the following way:
1. Create a file `input.json` with the following contents: `{"KeyId": "<MicrochipPartsTableKey>", "Plaintext": "<SecureElementKeyValueInBase64>"}`
2. `aws kms encrypt --cli-input-json file://input.json`. It is important to use the input json file so that secret value doesn't stay in history
3. Change the contents of `input.json` to `{"TableName": "<MicrochipPartsTableKey>", "Item": {"part_number": {"S": "<PartNumber>"}, "psk_root_keys_encrypted": {"S": "<CiphertextBlobFromKMSEncrypt>"}}}`
4. `aws dynamodb put-item --cli-input-json file://input.json`
5. `rm input.json`

At this point interoperability should be working. It might be required to restart the containers before they pick up new configuration.

# Identity Server interoperability

> Identity Server interoperability doesn't need Crypto Server, therefore you don't need to set up two accounts and peer networks. Still, you need to set up TLS certificates in a similar fasion.

Reviewer: ```suggestion
Identity Server interoperability doesn't need Crypto Server, therefore you don't need to set up two accounts and peer networks. Still, you need to set up TLS certificates in a similar fashion.
```

---
title: "Interoperability"
description: "LoRaWAN Backend Interfaces"
weight: 3
---

> {{% tts %}} offers support for LoRaWAN Backend Interfaces. Interoperability can be turned on on JoinServer or IdentityServer, but not both at once, as they use the same port.

Reviewer: ```suggestion
{{% tts %}} exposes the Join Server API defined by LoRaWAN Backend Interfaces 1.0 and 1.1. This is used for interoperability between LoRaWAN networks. The functionality can be exposed either by the Join Server or by the Identity Server, but not both at once.
```

> # JoinServer interoperability

Reviewer: ```suggestion
# Join Server interoperability
```

## Architecture

Support for LoRaWAN Backend Interfaces requires certain extensions to the typical AWS ECS deployment architecture. First, we need to include CryptoServer in the deployment. This component is responsible for cryptographic operations in the context of Microchip secure element. Since it needs access to highly sensitive data, {{% tts %}} assumes that it's not deployed together with the rest of the stack, but in a separate account. In this way regular ops team can have access to the most part of the stack, while CryptoServer, together with highly sensitive Microchip secure element encryption keys, would be accessed only by selected members of your organization.

> In order to provide this, {{% tts %}} uses VPC peering functionality of AWS.

Reviewer: Remove; this is not relevant for public documentation. Also, how things are operated in TTSC(E) are not documented publicly.

> Moreover, JoinServer needs to validate the client certificate when establishing TLS connections. This means that, while the rest of the stack has its TLS terminated by the AWS Load Balancer, JoinServer needs access to the TLS certificate of your server and of connecting clients.

Reviewer: Idem; too deployment specific
